File: No.2490-Bees-master/api/src/main/java/org/bees/biz/service/AdminService.java
package org.bees.biz.service;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.security.NoSuchAlgorithmException;
import java.util.LinkedList;
import java.util.List;

import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.apache.commons.beanutils.PropertyUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.bees.biz.persistence.manager.AdminManager;
import org.bees.biz.persistence.model.Admin;
import org.bees.biz.service.dto.AdminDto;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import projects.commons.exception.ServiceException;
import projects.commons.service.cache.CacheService;
import projects.commons.utils.MD5Utils;
import projects.commons.utils.RandomStringUtils;
import projects.commons.utils.ValidateUtils;
import projects.commons.utils.config.GlobalKeys;
import projects.commons.utils.response.Res;

@Service
public class AdminService {
	
	@Autowired
	private AdminManager adminManager;
	
	@Autowired
	private CacheService cacheService;
	
	
	@SuppressWarnings("unused")
	private final static Log LOG = LogFactory.getLog(AdminService.class);

	public Admin getInfoById(Integer id) {		
		return this.adminManager.getInfoById(id);
	}

	public List<AdminDto> getList(int start, int limit) {
		List<Admin> Admins=this.adminManager.getList(start,limit);
		List<AdminDto> dtos=new LinkedList<AdminDto>();
		for (Admin Admin : Admins) {
			AdminDto dto=new AdminDto();
			try {
				PropertyUtils.copyProperties(dto, Admin);
				
			} catch (IllegalAccessException | InvocationTargetException
					| NoSuchMethodException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			dtos.add(dto);
		}		
		return dtos;
	}

	public Res login(String name, String password, HttpServletRequest request,
			HttpServletResponse response) {
		Res res = new Res();
		try {
			Integer errorNum = cacheService
					.getAttribute(GlobalKeys.PASSPORT_LOGIN_ERROR_KEY + name,
							Integer.class);
			if (errorNum == null) {
				errorNum = 0;
			}
			if (errorNum >= 10) {
				throw new ServiceException("登录失败超过10次，请15分钟后重试");
			}
			
			Admin admin=this.adminManager.getInfoByName(name);
			if(ValidateUtils.isNull(admin)){
				throw new ServiceException("用户名或密码错误，请重新输入");
			}
			if (!admin.getPasswd().equals(
					this.getSafePass(password))) {
				cacheService.setAttribute(GlobalKeys.PASSPORT_LOGIN_ERROR_KEY
						+ name, errorNum + 1);
				cacheService.setExpire(GlobalKeys.PASSPORT_LOGIN_ERROR_KEY
						+ name, 900);// 15分钟过期
				throw new ServiceException("用户名或密码错误，请重新输入");
			}
			admin.setPasswd("");// 清除密码
			
	
			String sessionId = RandomStringUtils.randomAlphanumeric(32)
					.toUpperCase() + "_" + admin.getId()+"_ADMIN";
			cacheService.setAttribute(
					GlobalKeys.PASSPORT_LOGIN_KEY + sessionId, admin);
			
			res.addResponse("ops", admin.getOps());
			res.addResponse("sessionKey", sessionId);
			res.addResponse("flag", true);
			
			
			Cookie cookie = new Cookie(GlobalKeys.PASSPORT_LOGIN_KEY, sessionId);
			cookie.setMaxAge(-1);
			cookie.setPath("/");
			//cookie.setDomain(GlobalConfigure.AUTH_DOMAIN);
			response.addCookie(cookie);
			// res.setRedirect("index.html");
			res.commit();
			return res;
		} catch (ServiceException e) {
			throw e;
		} catch (Exception e) {			
			throw new ServiceException(e);
		}
		
	}

	
	public String getSafePass(String password) {
		try {
			return MD5Utils
					.getMD5String(MD5Utils
							.getMD5String(password) + GlobalKeys.MD5KEY);
		} catch (NoSuchAlgorithmException | IOException e) {
			
		}
		return password;
	}
	
	public Integer getUserId(String sessionKey) {
		if (!ValidateUtils.isNull(sessionKey)) {
			Admin admin = cacheService.getAttribute(GlobalKeys.PASSPORT_LOGIN_KEY
					+ sessionKey, Admin.class);
			if (admin != null&&isAdmin(sessionKey)) {
				return admin.getId();
			}
		}
		return null;
	}
	
	public boolean isAdmin(String sessionKey) {
		String[] keys=sessionKey.split("_");
		if(keys[keys.length-1].equals("ADMIN")){
			return true;
		}
		return false;
	}

	public void hasOp(Integer adminId, int op) {
		Admin admin=this.getInfoById(adminId);
		if(ValidateUtils.isNull(admin)){
			throw new ServiceException("用户错误");
		}
		boolean hasop=false;
		String[] ops=admin.getOps().split(",");
		for (String string : ops) {
			if(string.equals(op+"")){
				hasop=true;
				break;
			}
		}
		if(!hasop){
			throw new ServiceException("无权限");
		}
		
	}

	public int getListCount() {
		return this.adminManager.getListCount();
	}

	public boolean setPassword(Integer adminId, String oldPassword,
			String newPassword) {
		Admin admin=this.adminManager.getInfoById(adminId);
		if(!ValidateUtils.isNull(admin)&&admin.getPasswd().equals(this.getSafePass(oldPassword))){
			admin.setPasswd(this.getSafePass(newPassword));
			int res=this.adminManager.getDao().update(admin);
			if(res>0){
				return true;
			}else{
				return false;
			}
		}
		return false;
	}

	public boolean add(String name, String password, String ops) {
		password=this.getSafePass(password);
		
		return this.adminManager.add(name, password, ops);
	}
	
	public boolean eidt(int id,String name, String password, String ops) {
		if(!ValidateUtils.isNull(password)&&!password.equals("")){
			password=this.getSafePass(password);		
		}
		return this.adminManager.edit(id,name, password, ops);
	}

	public boolean del(Integer id) {
		
		return this.adminManager.del(id);
	}
}

File: beesInterface/bees_admin/adminService.py
#coding:utf-8
from models import Admin
from utils.res import Res
from utils.globalKeys import GlobalKeys
from exception.serviceException import ServiceException
from utils.validateUtils import ValidateUtils
from utils.cacheService import CacheService
from utils.randomStringUtils import RandomStringUtils
from utils.md5Utils import MD5Utils
from utils.const import Const

class AdminService:

	@staticmethod
	def getInfoById(id):
		try:
			return Admin.objects.get(id=id, del_flag=Const.DEL_FLAG_NOT_DEL)
		except Exception, e:
			return None

	@staticmethod
	def getList(start, limit):
		admins = Admin.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).order_by('-id')
		if ValidateUtils.isNull(start):
			start = 0
		else:
			start = int(start)
		if ValidateUtils.isNull(limit):
			admins = admins[start:]
		else:
			limit = int(limit)
			admins = admins[start: start+limit]
		return admins

	@staticmethod
	def login(name, password, request, response):
		res = Res()
		try:
			errorNum = CacheService.getAttribute(GlobalKeys.PASSPORT_LOGIN_ERROR_KEY + name)
			if errorNum == None:
				errorNum = 0
			if errorNum >= 10:
				raise ServiceException("登录失败超过10次,请15分钟后重试")
			admin = None
			try:
				admin = Admin.objects.get(name=name)
			except Exception, e:
				pass
			if ValidateUtils.isNull(admin):
				raise ServiceException("用户名或密码错误，请重新输入")
			if admin.passwd != AdminService.getSafePass(password):
				CacheService.setAttribute(GlobalKeys.PASSPORT_LOGIN_ERROR_KEY + name, errorNum + 1)
				CacheService.setExpire(GlobalKeys.PASSPORT_LOGIN_ERROR_KEY + name, 900)  # 15分钟过期
				raise ServiceException("用户名或密码错误，请重新输入")
			admin.passwd = "" # 清除密码
			sessionId = RandomStringUtils.randomAlphanumeric(32).upper() + "_" + str(admin.id) + "_ADMIN"
			CacheService.setAttribute(GlobalKeys.PASSPORT_LOGIN_KEY + sessionId, admin)
			res.addResponse("ops", admin.ops)
			res.addResponse("sessionKey", sessionId)
			res.addResponse("flag", True)
			response.set_cookie(GlobalKeys.PASSPORT_LOGIN_KEY, sessionId, max_age=-1, path='/')
			return res
		except ServiceException, e:
			raise e
		except Exception, e:
			raise ServiceException(e)

	@staticmethod
	def getSafePass(password):
		try:
			return MD5Utils.getMD5String(MD5Utils.getMD5String(password)+ GlobalKeys.MD5KEY)
		except ServiceException, e:
			raise Exception(e)
		return password

	@staticmethod
	def getUserId(sessionKey):
		if not ValidateUtils.isNull(sessionKey):
			admin = CacheService.getAttribute(GlobalKeys.PASSPORT_LOGIN_KEY + sessionKey)
			if admin and AdminService.isAdmin(sessionKey):
				return admin.id
		return None

	@staticmethod
	def isAdmin(sessionkey):
		keys = sessionkey.split("_")
		if keys[-1] == "ADMIN":
			return True
		return False

	@staticmethod
	def hasOp(adminId, op):
		admin = AdminService.getInfoById(adminId)
		if ValidateUtils.isNull(admin):
			raise ServiceException("用户错误")
		hasop = False
		ops = admin.ops.split(",")
		for string in ops:
			if string == str(op):
				hasop = True
				break
		if not hasop:
			raise ServiceException("无权限")

	@staticmethod
	def getListCount():
		return Admin.objects.filter(del_flag=Const.DEL_FLAG_NOT_DEL).count()

	@staticmethod
	def setPassword(adminId, oldPassword, newPassword):
		admin = AdminService.getInfoById(adminId)
		if not ValidateUtils.isNull(admin) and admin.passwd == AdminService.getSafePass(oldPassword):
			admin.passwd = AdminService.getSafePass(newPassword)
			admin.save()
			return True
		else:
			return False

	@staticmethod
	def add(name, password, ops):
		try:
			Admin.objects.get(name=name) # 如果存在则返回
			return False
		except Exception, e:
			password = AdminService.getSafePass(password)
			admin = Admin(name=name, passwd=password, ops=ops)
			admin.name = name
			admin.passwd = password
			admin.ops = ops
			# admin.add_time = datetime.now()
			admin.save()
			return True

	@staticmethod
	def eidt(id, name, password, ops):
		try:
			admin = Admin.objects.get(id=id)
			admin.name = name
			if not ValidateUtils.isNull(password):
				admin.passwd = AdminService.getSafePass(password)
			admin.ops = ops
			admin.save()
			return True
		except Exception, e:
			return False

	@staticmethod
	def delete(id):
		try:
			admin = Admin.objects.get(id=id)
			admin.del_flag = Const.DEL_FLAG_DEL
			admin.save()
			return True
		except Exception, e:
			return  False

